dmoj/cptbox/ptdebug.cpp

```diff
@@ -49,6 +53,50 @@ void pt_debugger::settid(pid_t tid) {
 }
 #endif
 
+bool pt_debugger::pre_syscall() {
+#if PTBOX_FREEBSD
+    if (ptrace(PT_GETREGS, tid, (caddr_t) &regs, 0)) {
+        perror("ptrace(PT_GETREGS)");
+#else
+    struct iovec iovec;
+    iovec.iov_base = &regs;
+    iovec.iov_len = sizeof regs;
+
+    if (ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iovec)) {
+        perror("ptrace(PTRACE_GETREGSET)");
+#endif
+        abi_ = PTBOX_ABI_INVALID;
+        return false;
+    } else {
+#if PTBOX_FREEBSD
+        abi_ = abi_from_reg_size(sizeof regs);
+#else
+        abi_ = abi_from_reg_size(iovec.iov_len);
+#endif
+        regs_changed = false;
+        return true;
+    }
+}
+
+bool pt_debugger::post_syscall() {
+    if (!regs_changed || abi_ == PTBOX_ABI_INVALID)
```

Comment:
```suggestion
    assert(abi_ != PTBOX_ABI_INVALID);
    if (!regs_changed)
```

+ a comment saying that this should already be handled?